File: 4.4.2023/ex9.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.*;


class Person {
    protected String name;
    protected String address;

    public Person(String name, String address) {
        this.name = name;
        this.address = address;
    }

    @Override
    public String toString() {
        return "Name: " + name + ", Address: " + address;
    }
}

class Student extends Person {
    private double score1;
    private double score2;

    public Student(String name, String address, double score1, double score2) {
        super(name, address);
        this.score1 = score1;
        this.score2 = score2;
    }

    public double calculateAverageScore() {
        return (score1 + score2) / 2;
    }

    public String evaluate() {
        double avgScore = calculateAverageScore();
        if (avgScore >= 8.0) {
            return "Excellent";
        } else if (avgScore >= 6.5) {
            return "Good";
        } else if (avgScore >= 5.0) {
            return "Average";
        } else {
            return "Poor";
        }
    }

    @Override
    public String toString() {
        return super.toString() + ", Average Score: " + calculateAverageScore() + ", Evaluation: " + evaluate();
    }
}

class Employee extends Person {
    private double salaryCoefficient;

    public Employee(String name, String address, double salaryCoefficient) {
        super(name, address);
        this.salaryCoefficient = salaryCoefficient;
    }

    public double calculateSalary() {
        return salaryCoefficient * 1000;
    }

    public String evaluate() {
        double salary = calculateSalary();
        if (salary >= 20000) {
            return "High salary";
        } else if (salary >= 12000) {
            return "Average salary";
        } else {
            return "Low salary";
        }
    }

    @Override
    public String toString() {
        return super.toString() + ", Salary: " + calculateSalary() + ", Evaluation: " + evaluate();
    }
}

class Customer extends Person {
    private String companyName;
    private double invoiceValue;

    public Customer(String name, String address, String companyName, double invoiceValue) {
        super(name, address);
        this.companyName = companyName;
        this.invoiceValue = invoiceValue;
    }

    public String evaluate() {
        if (invoiceValue >= 100000) {
            return "VIP";
        } else if (invoiceValue >= 50000) {
            return "Gold";
        } else if (invoiceValue >= 10000) {
            return "Silver";
        } else {
            return "Normal";
        }
    }

    @Override
    public String toString() {
        return super.toString() + ", Company: " + companyName + ", Invoice Value: " + invoiceValue + ", Evaluation: " + evaluate();
    }
}
class Management {
    private Person[] people;
    private int totalPeople;
    private int capacity;

    public Management(int capacity) {
        this.capacity = capacity;
        this.people = new Person[capacity];
        this.totalPeople = 0;
    }

    public void addPerson(Person person) {
        if (totalPeople >= capacity) {
            capacity = (int) (capacity * 1.5);
            people = Arrays.copyOf(people, capacity);
        }
        people[totalPeople++] = person;
    }

    public void removePersonByName(String name) {
        for (int i = 0; i < totalPeople; i++) {
            if (people[i].name.equals(name)) {
                for (int j = i; j < totalPeople - 1; j++) {
                    people[j] = people[j + 1];
                }
                totalPeople--;
                break;
            }
        }
    }

    public void sortByName() {
        Arrays.sort(people, 0, totalPeople, Comparator.comparing(person -> person.name));
    }

    public void printList() {
        for (int i = 0; i < totalPeople; i++) {
            System.out.println(people[i]);
        }
    }
}

public class ex9 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter the initial capacity:");
        int capacity = scanner.nextInt();
        Management management = new Management(capacity);

        while (true) {
            System.out.println("\nMenu:");
            System.out.println("1. Add a person");
            System.out.println("2. Remove a person by name");
            System.out.println("3. Sort the list by name");
            System.out.println("4. Print the list");
            System.out.println("5. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            scanner.nextLine(); 

            if (choice == 5) {
                break;
            }

            switch (choice) {
                case 1:
                    System.out.println("Enter the person type (Student, Employee, Customer):");
                    String type = scanner.nextLine();
                    System.out.println("Enter the name:");
                    String name = scanner.nextLine();
                    System.out.println("Enter the address:");
                    String address = scanner.nextLine();

                    if (type.equalsIgnoreCase("Student")) {
                        System.out.println("Enter the score for subject 1:");
                        double subject1Score = scanner.nextDouble();
                        System.out.println("Enter the score for subject 2:");
                        double subject2Score = scanner.nextDouble();
                        management.addPerson(new Student(name, address, subject1Score, subject2Score));
                    } else if (type.equalsIgnoreCase("Employee")) {
                        System.out.println("Enter the salary coefficient:");
                        double salaryCoefficient = scanner.nextDouble();
                        management.addPerson(new Employee(name, address, salaryCoefficient));
                    } else if (type.equalsIgnoreCase("Customer")) {
                        System.out.println("Enter the company name:");
                        String companyName = scanner.nextLine();
                        System.out.println("Enter the invoice value:");
                        double invoiceValue = scanner.nextDouble();
                        management.addPerson(new Customer(name, address, companyName, invoiceValue));
                    } else {
                        System.out.println("Invalid person type. Please try again.");
                    }
                    break;
                case 2:
                    System.out.println("Enter the name of the person to remove:");
                    String personName = scanner.nextLine();
                    management.removePersonByName(personName);
                    break;
                case 3:
                    management.sortByName();
                    break;
                case 4:
                    management.printList();
                    break;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }

        scanner.close();
    }
}

File: 4.4.2023/ex9.py
from dataclasses import dataclass
from typing import List
import functools

@dataclass
class Person:
    name: str
    address: str

    def __str__(self):
        return f"Name: {self.name}, Address: {self.address}"


class Student(Person):
    def __init__(self, name, address, score1, score2):
        super().__init__(name, address)
        self.score1 = score1
        self.score2 = score2

    def calculate_average_score(self):
        return (self.score1 + self.score2) / 2

    def evaluate(self):
        avg_score = self.calculate_average_score()
        if avg_score >= 8.0:
            return "Excellent"
        elif avg_score >= 6.5:
            return "Good"
        elif avg_score >= 5.0:
            return "Average"
        else:
            return "Poor"

    def __str__(self):
        return super().__str__() + f", Average Score: {self.calculate_average_score()}, Evaluation: {self.evaluate()}"


class Employee(Person):
    def __init__(self, name, address, salary_coefficient):
        super().__init__(name, address)
        self.salary_coefficient = salary_coefficient

    def calculate_salary(self):
        return self.salary_coefficient * 1000

    def evaluate(self):
        salary = self.calculate_salary()
        if salary >= 20000:
            return "High salary"
        elif salary >= 12000:
            return "Average salary"
        else:
            return "Low salary"

    def __str__(self):
        return super().__str__() + f", Salary: {self.calculate_salary()}, Evaluation: {self.evaluate()}"


class Customer(Person):
    def __init__(self, name, address, company_name, invoice_value):
        super().__init__(name, address)
        self.company_name = company_name
        self.invoice_value = invoice_value

    def evaluate(self):
        if self.invoice_value >= 100000:
            return "VIP"
        elif self.invoice_value >= 50000:
            return "Gold"
        elif self.invoice_value >= 10000:
            return "Silver"
        else:
            return "Normal"

    def __str__(self):
        return super().__str__() + f", Company: {self.company_name}, Invoice Value: {self.invoice_value}, Evaluation: {self.evaluate()}"

class Management:
    def __init__(self, capacity):
        self.capacity = capacity
        self.people = [None] * capacity
        self.total_people = 0

    def add_person(self, person):
        if self.total_people >= self.capacity:
            self.capacity = int(self.capacity * 1.5)
            self.people.extend([None] * (self.capacity - len(self.people)))
        self.people[self.total_people] = person
        self.total_people += 1

    def remove_person_by_name(self, name):
        for i in range(self.total_people):
            if self.people[i].name == name:
                del self.people[i]
                self.total_people -= 1
                break

    def sort_by_name(self):
        self.people = sorted(self.people[:self.total_people], key=lambda person: person.name)

    def print_list(self):
        for i in range(self.total_people):
            print(self.people[i])

def main():
    capacity = int(input("Enter the initial capacity:"))
    management = Management(capacity)

    while True:
        print("\nMenu:")
        print("1. Add a person")
        print("2. Remove a person by name")
        print("3. Sort the list by name")
        print("4. Print the list")
        print("5. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 5:
            break

        if choice == 1:
            person_type = input("Enter the person type (Student, Employee, Customer): ")
            name = input("Enter the name: ")
            address = input("Enter the address: ")

            if person_type.lower() == "student":
                score1 = float(input("Enter the score for subject 1: "))
                score2 = float(input("Enter the score for subject 2: "))
                management.add_person(Student(name, address, score1, score2))
            elif person_type.lower() == "employee":
                salary_coefficient = float(input("Enter the salary coefficient: "))
                management.add_person(Employee(name, address, salary_coefficient))
            elif person_type.lower() == "customer":
                company_name = input("Enter the company name: ")
                invoice_value = float(input("Enter the invoice value: "))
                management.add_person(Customer(name, address, company_name, invoice_value))
            else:
                print("Invalid person type. Please try again.")

        elif choice == 2:
            person_name = input("Enter the name of the person to remove: ")
            management.remove_person_by_name(person_name)

        elif choice == 3:
            management.sort_by_name()

        elif choice == 4:
            management.print_list()

        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()

